src/main/java/com/raisetech/mybatisdemo/service/UserService.java

```diff
@@ -0,0 +1,12 @@
+package com.raisetech.mybatisdemo.service;
+
+import com.raisetech.mybatisdemo.entity.User;
+import com.raisetech.mybatisdemo.form.CreateForm;
+
+import java.util.List;
+
+public interface UserService {
+    List<User> findAll();
+
+    User createUser(CreateForm form);
```

Comment:
業務ロジックがHTTPリクエストのフォームに依存しないような書き方をしたい。

```suggestion
    User createUser(String name, String address, Integer age);
```

Reply:
またはServiceに渡すときはDTOというものに変換する
```suggestion
    User createUser(CreateDto form);
```
※DTOも自前で作らないといけないですが……

Reply:
> 業務ロジックがHTTPリクエストのフォームに依存しないような書き方をしたい

ひとまず、String name, String address, Integer ageに修正しました。実務上では、DTOを使用することが多いのでしょうか？？

Reply:
> 業務ロジックがHTTPリクエストのフォームに依存しないような書き方をしたい。

フォームは値を受け取るだけの為に使い、データベースへの登録処理の際にはフォームを使うべきではないということですね！ただその際に、名前や住所や年齢はどのようにcreateUserに渡すのでしょうか💦フォームで受け取った値を、Userクラスに格納し、UserクラスからcreateUserに渡すということでしょうか・・？わかってなくてトンチンカンな質問になっていたらすみません。
DTOを作った場合は、DTOに値を格納して、createUserに渡すのかなとは思うのですが。

Reply:
> トンチンカンな質問になっていたらすみません。

そんなことないですよ！
いい質問です。

Reply:
たとえば、UserServiceのcreateUserメソッドのシグネチャをこのようにするとします。
※シグネチャについても調べてみてください 🚀 
```java
User createUser(String name, String address, Integer age);
```

このときにUserServieImplの実装は下記のようにすればよいです。
```java
@Override
public User createUser(String name, String address, Integer age) {
    User user = new User( form.getName(), form.getAddress(), form.getAge());
    userMapper.createUser(user);
    return user;
}
```
ポイントとしては
- UserクラスにIDを指定しなくてよいコンストラクタを作る
- Userのインスタンス生成時にはIDはnullまたはintのデフォルト値である0が設定される
- userMapper.createUserでデータ登録時にuserインスタンスにIDが設定される
参考: [MyBatis で Auto Increment された ID を取得する](https://autopp-tech.hatenablog.com/entry/2019/08/29/093508)

Reply:
上記に書いていただいてる内容を何度も確認しましたが、全然理解できていない状況です・・。
引数を「String name, String address, Integer age」に変更したところ、エラーの嵐となり解決できていません。
<img width="1440" alt="スクリーンショット 2023-06-11 16 57 42" src="https://github.com/kima-s/lecture09/assets/130459578/d3f9fef1-dabb-434a-957a-98cf7294809b">
この場合、UserServieImplファイルで「form.getName(), form.getAddress(), form.getAge()」メソッドを使用する為に、インスタンス化や依存性の注入をする必要があるということでしょうか？初歩的なところがわかっていないのかもしれません・・

Reply:
> ポイントとしては
UserクラスにIDを指定しなくてよいコンストラクタを作る
Userのインスタンス生成時にはIDはnullまたはintのデフォルト値である0が設定される
userMapper.createUserでデータ登録時にuserインスタンスにIDが設定される
参考: [MyBatis で Auto Increment された ID を取得する](https://autopp-tech.hatenablog.com/entry/2019/08/29/093508)

また、コメントいただいていた部分は「@Options(useGeneratedKeys=true, keyColumn="id")」を使用すれば良いということはわかりましたが、「UserクラスにIDを指定しなくてよいコンストラクタを作る」の部分をどの様にすれば良いのかがわかりません。何かヒントをいただけないでしょうか？

Reply:
対応ありがとうございます！

そして、お悩み確認しました。
ごめんなさい 🙇 
私の実装例が誤っていました。

```java
@Override
public User createUser(String name, String address, Integer age) {
    User user = new User( form.getName(), form.getAddress(), form.getAge());
```
ですが、ただしくは
```java
@Override
public User createUser(String name, String address, Integer age) {
    User user = new User(name, address, age);
```
です。
ServiceがFormに依存するのを排除してほしいといっていたのに……

---

次に、「UserクラスにIDを指定しなくてよいコンストラクタを作る」ですが、このようなイメージです。
```java
// 登録時に利用するコンストラクタ
// idはMｙBatisによって補完されるのでいったん0を代入しておく
public User(String name, String address, int age) {
    this.id = 0;
    this.name = name;
    this.address = address;
    this.age = age;
}
```

Reply:
そういうことだったのですね！！Formには依存しないとの話だったのに、私もどの様にすればメソッドを持って来れるかずっと考えていて気づきませんでした・・。
上記の様に入力してみました。ただ、userControllerの一部だけはformを使用してしまいました。問題ありませんでしょうか？
<img width="1440" alt="スクリーンショット 2023-06-11 21 31 30" src="https://github.com/kima-s/lecture09/assets/130459578/608a342d-c077-4e57-861f-fa76c06802eb">

Reply:
また、「@options(useGeneratedKeys=true, keyColumn="id")」を使用しましたが、postmanのURLにget.Idが反映されていません。デバッグも試してみましたが、やり方が悪いのかなぜ反映されるのかわかりませんでした。
<img width="1440" alt="スクリーンショット 2023-06-11 21 30 58" src="https://github.com/kima-s/lecture09/assets/130459578/64a265a2-adf4-4f9c-8bae-5afdfbd94712">
アドバイスを頂けると幸いです・・。本来、８などが表示されるはずなのですが・・